File: src/utils/get-faa-images.ts
// do a fetch to the faa images api and return the images
import axios from "axios";
import fs from "fs";
import path from "path";
import dotenv from "dotenv";
import { CamDataType } from "../types/cameraData";
dotenv.config();

const FAA_API_URL = "https://weathercams.faa.gov/api/redistributable/sites";
const HEADERS = {
  Authorization: `Bearer ${process.env.FAA_API_KEY}`,
};
const getFaaRawData = async () => {
  const res = await axios.get(FAA_API_URL, {
    headers: HEADERS,
  });
  const sites = res?.data?.payload;
  return sites as Promise<any[]>;
};

const getFaaImages = async () => {
  const sites = await getFaaRawData();
  const faaImages = sites
    .map((site) => {
      if (!site.cameras) return [];
      const cams: CamDataType[] = site?.cameras.map((camera: any) => {
        return {
          cameraName: `${site.siteName} ${camera.cameraDirection} ${
            site.operatedBy || ""
          }`,
          url: camera.currentImageUri,
          source: "faa.gov",
          lat: site.latitude,
          long: site.longitude,
          refreshRate: "10 min",
        };
      });
      return cams || [];
    })
    .flat();
  return faaImages;
};

export default getFaaImages;

File: python/utils/get_faa_images.py
import os
import requests
from dotenv import load_dotenv

load_dotenv()

FAA_API_URL = "https://weathercams.faa.gov/api/redistributable/sites"
FAA_API_KEY = os.getenv("FAA_API_KEY")

def get_faa_images():
    headers = {"Authorization": f"Bearer {FAA_API_KEY}"}
    response = requests.get(FAA_API_URL, headers=headers)
    response.raise_for_status()
    sites = response.json().get("payload", [])
    
    faa_images = []
    for site in sites:
        if not site.get("cameras"):
            continue
        for camera in site["cameras"]:
            faa_images.append({
                "cameraName": f"{site.get('siteName', '')} {camera.get('cameraDirection', '')} {site.get('operatedBy', '')}",
                "url": camera.get("currentImageUri"),
                "source": "faa.gov",
                "lat": site.get("latitude"),
                "long": site.get("longitude"),
                "refreshRate": "10 min",
                "camera_id": camera.get("cameraId"),
                "attribution": site.get("attribution", "FAA")
            })
    faa_images = [img for img in faa_images if img != None]
    faa_images = [img for img in faa_images if "Alert California" not in (img.get("attribution") or "")]
    return faa_images